### PR TITLE
AIP-44 Migrate Dagbag.sync_to_db to internal API.

### LINKED ISSUE #28615: AIP-44 Migrate Dagbag.sync_to_db to internal API.
This method is used in DagFileProcessor.process_file - it may be easier to migrate all it's internal calls instead of the whole method.

### PR DESCRIPTION
**Pull request by snjypl**
Fixes: [#28615](https://github.com/apache/airflow/issues/28615)

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by snjypl** [comment]
Hi Team  @potiuk  @mhenc  @mhenc 

this task appears to be more complex than i expected. i went through the [AIP 44, ](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-44+Airflow+Internal+API), and the PRs for the other tasks to get a better understanding. 

in AIP-44, i found a reference to the `save_dag_to_db` method.
> This is a new, “heaviest” method that should be exposed by the internal API to the DAG processor. We should be able to serialize all the information stored currently in the DagBag and send it to the Internal API server. The method should be roughly equivalent to: sync_to_db and pickling dags if set. This should be done in a single transaction as a single API call.

based on my understanding so far, i have done a rough refactoring to see if i am in the right direction. it will be really helpful if anyone could take a look at the changes. it is just a rough changes, i wanted to see if i understand the task correctly. 

from the AIP, i got the impression that we would be sending the serialized dag to the internal api.   will `BaseSerialization.serialize` in the `internal_api_call` decorator take care about the serialization?

**Comment by snjypl** [comment]
i tried the below code to check the serialization/deserialization:  
```
from airflow.models.dagbag import DagBag
from airflow.serialization.serialized_objects import BaseSerialization,SerializedDAG
dagbag = DagBag('/opt/airflow/dags')
BaseSerialization.deserialize(BaseSerialization.serialize(dagbag.dags))
```
but i got this error: 
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 475, in deserialize
    return {k: cls.deserialize(v) for k, v in var.items()}
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 475, in <dictcomp>
    return {k: cls.deserialize(v) for k, v in var.items()}
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 471, in deserialize
    var = encoded_var[Encoding.VAR]
KeyError: <Encoding.VAR: '__var'>
```

File: airflow/dag_processing/processor.py [hunk 1]
```diff
@@ -837,3 +832,26 @@ def process_file(
             self.log.exception("Error logging DAG warnings.")
 
         return len(dagbag.dags), len(dagbag.import_errors)
+
+    @staticmethod
+    @internal_api_call
+    @provide_session
+    def save_dag_to_db(
```
Comment (mhenc) [inline airflow/dag_processing/processor.py:837]: Please add types.

**Comment by mhenc** [comment]
> From the AIP, i got the impression that we would be sending the serialized dag to the internal api. Will `BaseSerialization.serialize` in the `internal_api_call` decorator take care about the serialization?

Yes, the InternalApi decorator/server serializes/deserializes the objects using BaseSerialization.

https://github.com/apache/airflow/blob/main/airflow/api_internal/internal_api_call.py#L107

 I see that you have problems serializing `dags`, which is `dict[String,DAG]`. We may need to extend the BaseSerialization to support it. Let me take a look.

File: airflow/dag_processing/processor.py (hunk 1 above)
Comment (snjypl) [reply airflow/dag_processing/processor.py:837]: > Please add types.
done

**Review by vincbeck — APPROVED** [review]

**Comment by potiuk** [comment]
Sorry, i have not looked at this before, I want to have a closer look at that after we discuss the approach i proposed in https://github.com/apache/airflow/pull/29513#discussion_r1106553896. - I have a feeling that once we agree this is a good idea and we apply it here, we might implement some simple refactor that will let us avoid any complexity (but it's too late for me now to think clearly).

**Review by potiuk — APPROVED** [review]
LGTM

**Pull request merged**